js/BranchingSet.js

```diff
@@ -0,0 +1,188 @@
+    const model = this.getNextModel();
+    this.addNextModel(model, true, false); // BUG?: check change to flag works
+  }
+
+  getNextModel() {
+    const config = this.model.get('_branching');
+    const brachingModels = this.models;
+    const branchedModels = this.branchedModels;
+
+    const isBeforeStart = !branchedModels.length;
+    if (isBeforeStart) {
+      const hasStartId = Boolean(config._start);
+      let firstModel;
+      if (hasStartId) firstModel = brachingModels.find(model => model.get('_id') === config._start);
+      else firstModel = brachingModels[0];
+      return firstModel;
+    }
+
+    const lastChildModel = branchedModels[branchedModels.length - 1];
+    const isLastIncomplete = !lastChildModel.get('_isComplete');
+    if (isLastIncomplete) {
+      return false;
+    }
+
+    const lastChildConfig = lastChildModel.get('_branching');
+    if (!lastChildConfig || !lastChildConfig._isEnabled) return true;
+
+    // Branch from the last model's correctness, if configured
+    const correctness = getCorrectness(lastChildModel);
+    const nextId = lastChildConfig['_force'] || lastChildConfig[`_${correctness}`];
+    if (!nextId) return true;
+
+    const isRelativeId = nextId.includes('@');
+    if (!isRelativeId) {
+      return brachingModels.find(model => model.get('_id') === nextId) || true;
+    }
+    
+    const originalLastChildModel = Adapt.findById(lastChildModel.get('_branchOriginalModelId'));
+    const nextModel = originalLastChildModel.findRelativeModel(nextId);
+    const wasModelAlreadyUsed = nextModel.get('_isAvailable');
+    if (wasModelAlreadyUsed) return true;
+    return nextModel;
+  }
+
+  addNextModel(nextModel, shouldSave = true, shouldRestore = false) {
+    // Clear the set's current back() position
+    this.model.set({
+      _branchLastPreviousIndex: null
+    });
+    // Increment the original model's attempts
+    const attemptIndex = (nextModel.get('_branchAttempts') || 0);
+    nextModel.set('_branchAttempts', attemptIndex + 1);
+    let wasAnyPartRestored = false;
+    const cloned = nextModel.deepClone((clone, model) => {
+      clone.set('_isAvailable', true);
+      // Remove tracking ids as these will change depending on the branches
+      // Clone attempt states are stored on the original model in their order of occurance
+      if (clone.has('_trackingId')) {
+        clone.unset('_trackingId');
+      }
+      if (clone instanceof QuestionModel) {
+        // Clear clone response history
+        clone.set('_attemptStates', false);
+        if (shouldRestore) {
+          // Restore the attempt state from the original model in occurance order
+          const attemptObjects = model.getAttemptObjects();
+          if (attemptObjects[attemptIndex]) {
+            clone.set(attemptObjects[attemptIndex]);
+            wasAnyPartRestored = true;
+          }
+          return;
+        }
+        // Reset if not restored or not a question
+        clone.reset('hard', true);
```

Comment:
The `not a question` part of the comment here implies that the reset should be outside of the if statement?

Reply:
Correct